File: Stage/Python/Core/Parametres.py
from Core.Element import C_Element

class C_Parametres:

    def __init__(self,xmlFile):
        self._parametres = {}
        self._vues = []
        listParametres = xmlFile.searchXml("Parametres")
        for parametre in listParametres:
            nom = parametre.get("nom")
            couleur = parametre.get("couleur")

            newParametre = C_Element(nom,couleur,"parametre")
            
            self._parametres[nom]= newParametre

    def abonne(self, vue):
        trouver = 0
        for v in self._vues:
            if (v == vue):
                trouver = 1

        if (trouver==0):
            self._vues.append(vue)

    def getListParametre(self):
        listParametres = []
        parametres = list(sorted(self._parametres.keys()))
        for parametre in parametres:
            listParametres.append(self._parametres.get(parametre))
            
        return listParametres

    def getListParametre_Noms(self, listNom):
        listeParametres = []
        for nom in listNom:
            listeParametres.append(self.getParametre(nom))
        return listeParametres

    def getParametre(self, nom):
        return self._parametres.get(nom)

    def getListNomParametre(self):
        return list(sorted(self._parametres.keys()))

    def getListCouleur(self):
        return list(self._parametres.values())

    def getCouleurParametre(self, nom):
        parametre = self._parametres.get(nom)
        return parametre.getCouleur()
        
    def setCouleurParametre(self, nom , couleur):
        parametre = self._parametres.get(nom)
        parametre.setCouleur(couleur)
        
        self.update()

    def setNomParametre(self, nomAvant, nomChanger):
        parametre = self._parametres.get(nomAvant)
        parametre.setNom(nomChanger)

        self._parametres[nomChanger] = self._parametres.pop(nomAvant)
        
        self.update()
        
    def update(self):
        for vue in self._vues:
            vue.refresh()

File: Stage/C++/Core/parametres.cpp
#include "parametres.h"

// classe qui permet de gerer tous les parametres
// herite de la classe modele a chaque modification d'un groupe la classe va appeler la fonction refresh de ces vues(qui sont les interfaces)
// parametres est un dictionnaire qui a pour cle le nom du parametre et comme valeur le pointeur vers le parametre

/*--------------------------------CONSTRUCTEURS--------------------------------*/
Parametres::Parametres(Xml *xml){
    QList<QDomElement> listElement = xml->getCible("Parametres");

    foreach(QDomElement element, listElement){
        QString nom = element.attribute("nom");
        QString couleur = element.attribute("couleur");
        QString categorie = element.attribute("categorie");
        Parametre *parametre = new Parametre(nom,couleur);
        parametre->setCategorie(categorie);
        parametres.insert(nom,parametre);
    }
}

/*--------------------------------DESTRUCTEURS--------------------------------*/
Parametres::~Parametres(){
    foreach(Parametre* elt, parametres.values()){
        if(elt!=NULL)
            delete elt;
    }
    parametres.clear();
}


/*--------------------------------FONCTIONS PUBLIC--------------------------------*/
QList<Element*> Parametres::getListParametre(){
    QList<Element*> listParametre;
    foreach(Element* parametre, parametres.values()){
        listParametre.append(parametre);
    }

    return listParametre;
}

// fonction qui peremt de recuperer une liste de parametres en fonction d'une liste de nom
QList<Element*> Parametres::getListParametre(QList<QString> listNom){
    QList<Element*> listParametre;
    foreach(QString nom, listNom){
        listParametre.append(parametres.value(nom));
    }

    return listParametre;
}

Element *Parametres::getParametre(QString nom){
    return parametres.value(nom);
}

QStringList Parametres::getListNomParametre(){
    return parametres.keys();
}

QStringList Parametres::getListCouleurParametre(){
    QStringList listCouleur;
    foreach(Parametre* parametre , parametres.values()){
        listCouleur.append(parametre->getCouleur());
    }
    return listCouleur;
}

QString Parametres::getCouleurParametre(QString nom){
    return parametres.value(nom)->getCouleur();
}

void Parametres::setCouleurParametre(QString nom, QString couleur){
    parametres.value(nom)->setCouleur(couleur);
    update();
}

void Parametres::setNomParametre(QString oldNom, QString newNom){
    Parametre* parametre = parametres.value(oldNom);
    parametre->setNom(newNom);
    parametres.remove(oldNom);
    parametres.insert(parametres.end(),newNom,parametre);
    update();
}

// fonction qui permet d'ajouter un parametre
void Parametres::addParametre(QString nom, QString unite){
    Parametre *parametre = new Parametre(nom,"0,0,0",unite);
    parametre->setCategorie("Nouveaux");
    parametres.insert(nom,parametre);
    update();
}

void Parametres::setCategorie(QString nom, QString categorie){
    Parametre *parametre = parametres.value(nom);
    parametre->setCategorie(categorie);
    update();
}

void Parametres::removeParametre(QString nomParametre){
    Parametre *parametre = parametres.value(nomParametre);
    parametres.remove(nomParametre);
    delete parametre;
    update();
}
